src/java/org/apache/cassandra/transport/messages/BatchMessage.java

```diff
@@ -223,20 +227,47 @@ protected Message.Response execute(QueryState state, long queryStartNanoTime, bo
             BatchStatement batch = new BatchStatement(null, batchType,
                                                       VariableSpecifications.empty(), statements, Attributes.none());
 
-            long queryTime = System.currentTimeMillis();
-            Message.Response response = handler.processBatch(batch, state, batchOptions, getCustomPayload(), queryStartNanoTime);
+            long requestStartMillisTime = System.currentTimeMillis();
+            Optional<Stage> asyncStage = Stage.fromStatement(batch);
+            if (asyncStage.isPresent())
+            {
+                BatchStatement finalStatement = batch;
+                List<QueryHandler.Prepared> finalPrepared = prepared;
+                return CompletableFuture.supplyAsync(() -> handleRequest(state, queryStartNanoTime, handler, batch, batchOptions, queries, statements, finalPrepared, requestStartMillisTime),
```

Comment:
use `asyncStage.get().submit` instead?
or do you wanted to emphasize it's async?

to me it sounds safer to use `submit`, in case it does something additional, internal to the stage. OTOH, if we have an API to get the `executor` it's not really realistic to keep it's usages in sync with any hanky-panky that may happen in `submit` et. al....

OK, so I don't have a strong opinion here.

Reply:
> use asyncStage.get().submit instead?

I just liked sticking to the `CompletableFuture` APIs to drive the async execution, if no strong reasons/opinions I'd just leave it like this.